lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp

```diff
@@ -0,0 +1,310 @@
+      assert(state.yieldState != BBState::BeforeYield);
+      return YieldError(MultipleYield, yield, state);
+    }
+
+    static YieldError getReturnBeforeYieldError(ReturnInst *returnI,
+                                                BBState state) {
+      assert(state.yieldState == BBState::BeforeYield);
+      return YieldError(ReturnBeforeYield, returnI, state);
+    }
+
+    static YieldError getReturnOnConflict(ReturnInst *returnI, BBState state) {
+      assert(state.yieldState == BBState::Conflict);
+      return YieldError(ReturnOnConflict, returnI, state);
+    }
+  };
+
+  /// Transfer function of the data-flow analysis.
+  ///
+  /// \param bb Basic block that should be processed
+  /// \param inState BBState at the start of the basic block
+  /// \param error out parameter that will contain information about
+  /// an error that is detected.
+  /// \return the state at the exit of the basic block if it can be computed
+  /// and None otherwise.
+  static Optional<BBState>
+  transferStateThroughBasicBlock(SILBasicBlock *bb, BBState inState,
+                                 Optional<YieldError> &error) {
+    auto *term = bb->getTerminator();
+
+    if (auto *returnInst = dyn_cast<ReturnInst>(term)) {
+      if (inState.yieldState == BBState::BeforeYield) {
+        error = YieldError::getReturnBeforeYieldError(returnInst, inState);
+        return None;
+      }
+
+      if (inState.yieldState == BBState::Conflict) {
+        error = YieldError::getReturnOnConflict(returnInst, inState);
+        return None;
+      }
+      return inState;
+    }
+
+    if (auto *yieldInst = dyn_cast<YieldInst>(term)) {
+      if (inState.yieldState != BBState::BeforeYield) {
+        error = YieldError::getMultipleYieldError(yieldInst, inState);
+        return None;
+      }
+
+      // If the current state is BeforeYield and if the basic block ends in a
+      // yield the new state is AfterYield.
+      return inState.getAfterYieldState(term);
+    }
+
+    // We cannot have throws within generalized accessors.
+    assert(!isa<ThrowInst>(term));
+
+    return inState;
+  }
+
+  /// Merge operation of the data-flow analysis.
+  ///
+  /// \param mergeBlock the basic block that is reached with two states
+  /// \param oldState the previous state at the entry of the basic block
+  /// \param newState the current state at the entry of the basic block
+  /// \return the new state obtained by merging the oldState with the newState
+  static BBState merge(SILBasicBlock *mergeBlock, BBState oldState,
+                       BBState newState) {
+    if (oldState.yieldState == BBState::Conflict) {
+      return oldState;
+    }
+
+    if (newState.yieldState == BBState::Conflict) {
+      return newState;
+    }
+
+    if (oldState.yieldState == newState.yieldState) {
+      return oldState;
+    }
+
+    // Here, one state is AfterYield and the other one is BeforeYield.
```

Comment:
Can you add an assertion for this? That way if the lattice changes but the merge function doesn't we'll get an assertion failure.